# 4.5 Practice Exercise 

In [1]:
#importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#import orders dataframe
df_ords=pd.read_csv(r'/Users/ritviksolanki/Desktop/Instacart Basket Analysis /02 Data/Original data/orders.csv',index_col=False)

In [3]:
#importing products Dataframe
df_prods=pd.read_csv(r'/Users/ritviksolanki/Desktop/Instacart Basket Analysis /02 Data/Original data/products.csv',index_col=False)

Question 2: Check df_ords dataframe. Share whether anything about the data looks off or should be investigated further.

In [4]:
#checking frequency of df_ords dataframe 
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

When looking at our frequency table for orders, it seems that Days since prior order has a minimum of 0, which makes sense. A maximum of 3 which also makes sense since there are 7 days in a week. However, it does not make sense that the 50th percentile value is 7 which is larger than the max.

In the number of orders we have a maximum of 100 and a minimum of 1. The 50th percentile is 5. this could make sense if most of the order numbers were very small.  

Question 3: Check df_ords for mixed datatypes

In [5]:
# checking for mixed data types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)


Since nothing was returned, there are NO mixed data types in the df_ords dataframe 


Question 5: Run a check for missing values in dataframe

In [6]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [7]:
# creating a new dataframe for missing values of days since prior order
df_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [8]:
#view the new dataframe df_nan
df_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [9]:
#searching the number_of_orders column for missing values for each value in our column.
df_ords['order_number'].value_counts(dropna=False) 

1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: order_number, Length: 100, dtype: int64

There are large numbers of missing values in days_since_prior_order of 206209. In each column, it seems that NaN correlates with number of orders=1. Tihs could make sense because when a user_id has a number of orders=1 they have not had any prior orders so days_since_prior_orders does not apply to those without prior orders.

Question 6: Address the missing values using an appropriate method. 

Since this missing value actually makes sense within the context of the data. A customer who only has one past order will have missing values for "days_since_prior_order". No need to remve these missing values in this case.

Question 7: Run a check for duplicate values in your df_ords data. 

In [10]:
# checking for duplicate values in df_ords
df_ords_dup=df_ords[df_ords.duplicated()]

In [11]:
# viewing duplicate records 
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

Answer 7: it appears that there are no duplicate records.

Question 8: address the duplicate records 

Answer 8: there are no duplicate records 

Question 9: Upload the dataframe df_ords and df_prods

In [12]:
path=r'/Users/ritviksolanki/Desktop/Instacart Basket Analysis /02 Data/Prepared Data'

In [13]:
#exporting df_ords 
df_ords.to_csv(os.path.join(path,'orders_checked.csv'))

In [14]:
#exporting df_prods
df_prods.to_csv(os.path.join(path,'product_checked.csv'))

In [16]:
df_ords.shape

(3421083, 7)

In [17]:
df_prods.shape

(49693, 5)

In [18]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0